shared/README.md

```diff
@@ -136,26 +137,119 @@ You'll get various prompts about installing various tools. You should
 install 'Build Tools' and any missing platforms. However, _don't_
 update the Android Gradle Plugin to 3.0.1.
 
+Also see below for some messages you may encounter with Android Studio.
+
 Some instructions talk about the SDK Manager / AVD Manager. This is
-under the Tools > Android menu. You may have to wait for the Gradle to
+under the Tools > Android menu. You may have to wait for Gradle to
 sync before they appear.
 
 To run on the Android simulator:
 
 ```sh
 # Build the go keybaselib
-yarn run rn-gobuild-android
-# if this fails with something like 'xcrun: error: unable to lookup item 'Path' in SDK 'iphoneos' you might have a partial xcode install. try
+yarn rn-gobuild-android
+
+# if this fails with something like "xcrun: error: unable to lookup item
+# 'Path' in SDK 'iphoneos'" you might have a partial xcode install. try
 xcode-select --switch /Applications/Xcode.app/Contents/Developer
-```
 
-```sh
 # Start the react native publisher (unlike on iOS, have to do this manually).
-yarn run rn-start
+yarn rn-start android
 ```
 
 Then do "Build > Make Project" and then "Run > Run 'app'".
 
+#### Dealing with Android Studio
+
+Sometimes Android Studio gets into a bad state, especially if you're
+doing stuff like `yarn modules` in the background.
+
+If 'Gradle sync' fails, you can retry it from Tools > Android > 'Sync
+Project with Gradle Files'.
+
+Sometimes, especially after opening Android Studio after a run of
+`yarn modules`, you'll get an "Unsupported Modules Detected" message
+for "react-native-fetch-blob", "react-native-contacts", and
+"react-native-image-picker". This seems to be harmless.
+
+Sometimes you'll also get an "An IDE Error has occured" message. That
+also seems to be harmless, although you may want to resync/reopen the
+project in that case.
+
+If nothing above works, you can try closing (File > Close Project) and
+reopening the project. Or even closing and reopening the app.
+
+#### Building and developing without Android Studio
+
+Alternatively, you can build and develop without Android
+Studio. However, first make sure you _can_ build and run with Android
+Studio first, as it's easier to get that working first.
+
+So make sure you've run `yarn rn-gobuild-android` and you have the
+react-native packager running (`yarn rn-start android`).
+
+Then run
+
+```sh
+# Build the apk.
+yarn rn-build-android
+```
+
+Unless you're modifying the Java files, you likely have to only run
+this occasionally.
+
+Then make sure you either have an emulator running, or you have your
+Android device connected (but not both). To check, run
+
+```sh
+adb devices
+```
+
+It should list exactly one device.
+
+To run the emulator, do:
+
+```sh
+cd $ANDROID_HOME/emulator
+emulator -list-avds
```

Comment:
Should we include adding `$ANDROID_HOME/tools` to `PATH` as a prerequisite?

Reply:
And `$ANDROID_HOME/platform-tools`?

Reply:
Ah, they're part of the react-native setup instructions, but added a comment mentioning that.